internal/server/main.go

```
@@ -0,0 +1,268 @@
+// Copyright 2022 The Go Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// server package implements job manager grpc server.
+package server
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	pb "github.com/alexbrainman/job/internal/proto"
+)
+
+// TODO: processOutput implementation does not support concurrent reading of stdout of the same job from multiple clients simulataniously
+
+// processOutput is used by process struct to implement
+// streaming of stdout and stderr.
+type processOutput struct {
+	stdstream io.Reader
+	readData  [][]byte
+}
+
+// output implements streaming of stdout and stderr of a process.
+func (po *processOutput) output(stream pb.JobManager_OutputServer) error {
+	// send data that we already read from stdout or stderr first
+	for _, buf := range po.readData {
+		err := stream.Send(&pb.Buffer{Bytes: buf})
+		if err != nil {
+			return err
+		}
+	}
+
+	// now try reading data from stdout or stderr
+	buf := make([]byte, 100)
+	for {
+		// read from stdout or stderr
+		n, err := po.stdstream.Read(buf)
+		if n == 0 {
```

It looks like you are not handling the case where `n > 0 && err != nil`

---

The client is going to receive a partial result when a non-EOF error occurs. It doesn't really matter if they receive more of a partial result. Failing fast seems like the appropriate move here.

---

I was suggesting that you could change from `n == 0` to `err != nil` in the code highlighted above.